### PR TITLE
docker: add "Security" tab to image listing

### PR DESCRIPTION
**Pull request by larskarlitski**
Add a "Security" tab to the image listing, if the "org.atomic" D-Bus service exists and that particular image has been scanned already.

You cannot initiate a scan from cockpit, yet (the "Scan again" button that doesn't work).

To try it out, install atomic and a scanner, for example `atomic install fedora/atomic_scan_openscap` and run `atomic scan [image]`. The security tab should then show up for that image.

### CODE REVIEW

**Comment by petems** [comment]
Awesome! Looking forward to container scanning showing up in Cockpit! 💯

**Comment by larskarlitski** [comment]
Pushed another update. The module now polls the `org.atomic` service every ten seconds for updated vulnerability information. We can change the timeout to something higher if that makes sense and remove it completely once [atomic issue #740](https://github.com/projectatomic/atomic/issues/740) is fixed.

I've removed the "Scan again" functionality until atomic supports it (filed under the same issue).

Also, the current `atomic_dbus.py` immediately drops the `org.atomic` name, but stays running. This means that cockpit is indirectly spawning lots of those processes while polling for updates. I reported it as [atomic issue #763](https://github.com/projectatomic/atomic/issues/763) and my fix landed in master recently.

**Comment by stefwalter** [comment]
Is this done? If so removing WIP would help get someone to review it.

**Comment by larskarlitski** [comment]
> Is this done? If so removing WIP would help get someone to review it.

Yes, I think this is a good first step that we can already include.

There are still some test failures though, so I'm leaving needswork.

**Comment by larskarlitski** [comment]
Rebased and fixed a whole bunch of edge cases (when atomic isn't installed or gives out inconsistent scan data). Also the `check-docker` tests that were failing should be ok again.

Also added a `warning` prop to `ListingRow`. Requested @dperpeet as a reviewer for that.

This is ready for review, so I'm removing needswork. I still need to write tests, but I'm not sure if it's ok to download the openscap container for each OS we test on.

**Comment by stefwalter** [comment]
> This is ready for review, so I'm removing needswork. I still need to write tests, but I'm not sure if it's ok to download the openscap container for each OS we test on.

Lets test it on the atomics for now ... does that work?

File: pkg/docker/containers-view.jsx [hunk 1]
```diff
@@ -400,13 +454,29 @@ var ImageList = React.createClass({
...
+            tabs.push({
+                name: _('Details'),
+                renderer: ImageDetails,
+                data: { image: image }
+            });
+
+            var vulnerable = false;
+            var vulnerableInfo = this.state.vulnerableInfos[image.Id.replace(/^sha256:/, '')];
+            if (vulnerableInfo !== undefined) {
+                vulnerable = true; // vulnerableInfo.vulnerabilities.length;
```
Comment (dperpeet) [inline pkg/docker/containers-view.jsx]: I think this is a debugging change.

**Comment by dperpeet** [comment]
Would it be a lot of work to move the changes to the listing component into a separate commit?

**Comment by dperpeet** [comment]
Let's add an integration test also, I think testing it on the Atomics like stef suggested would be sufficient for now.

**Comment by dperpeet** [comment]
> Would it be a lot of work to move the changes to the listing component into a separate commit?

Also, these changes should be reflected on the playground page.

**Comment by larskarlitski** [comment]
Updated.

> I think this is a debugging change.

Thanks for the catch!

> Would it be a lot of work to move the changes to the listing component into a separate commit?

I've removed them in favor of a separate column with a badge and vulnerability count, as we discussed on Friday.

> Let's add an integration test also, I think testing it on the Atomics like stef suggested would be sufficient for now.

I started this, but atomic's D-Bus service is broken on fedora-atomic. `python-slip-dbus` is not installed.

File: pkg/docker/atomic.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                    cockpit.all(promises).done(function () {
+                        var detailedInfos = {};
+
+                        for (var i = 0; i < arguments.length; i++)
+                            detailedInfos[ids[i]] = sanitizeVulnerableInfo(ids[i], arguments[i]);
+
+                        atomic.dispatchEvent("vulnerableInfoChanged", detailedInfos);
+                    });
+                }
+            } catch (error) {
```
Comment (stefwalter) [inline pkg/docker/atomic.js]: This isn't okay. If there are errors to ignore ... it should be around a single specific line. For example if we want to ignore ```JSON.parse()``` errors then we shuold wrap just that line in a ```try/catch``` handler.

File: pkg/docker/containers-view.jsx [hunk 3]
```diff
@@ -290,7 +291,7 @@ var ImageDetails = React.createClass({
         if (image.Config) {
             entrypoint = image.Config.Entrypoint;
             command = image.Config.Cmd;
-            ports = image.Config.ExposedPorts || [];
+            ports = Object.keys(image.Config.ExposedPorts || {});
```
Comment (stefwalter) [inline pkg/docker/containers-view.jsx]: Is this change logically separate? It should be a different commit if so.

File: pkg/docker/containers-view.jsx [hunk 4]
```diff
@@ -371,60 +417,98 @@ var ImageList = React.createClass({
...
-        var filtered = this.state.images.filter(function (image) {
-            return (image.RepoTags &&
-                    image.RepoTags[0].toLowerCase().indexOf(this.props.filterText.toLowerCase()) >= 0);
-        }.bind(this));
+    renderRow: function (image, withVul) {
+        var vulnerabilityColumn = '';
+
+        var vulnerableInfo = this.state.vulnerableInfos[image.Id.replace(/^sha256:/, '')];
+        if (vulnerableInfo) {
+            var count = vulnerableInfo.vulnerabilities.length;
```
Comment (stefwalter) [inline pkg/docker/containers-view.jsx]: Variable hoisting is at play here. You can either use ```let``` or declare ```count``` at the function level.

File: pkg/docker/containers-view.jsx [hunk 5]
```diff
@@ -308,6 +309,50 @@ var ImageDetails = React.createClass({
...
+
+        if (info.successful === false) {
+            text = cockpit.format(_('The scan from $time ($type) was not successful.'), args);
+
+        } else if (args.count === 0) {
+            text = cockpit.format(_('The scan from $time ($type) found no vulnerabilities.'), args);
+
+        } else {
+            text = cockpit.ngettext(cockpit.format('The scan from $time ($type) found one vulnerability:', args),
+                                    cockpit.format('The scan from $time ($type) found $count vulnerabilities:', args),
```
Comment (stefwalter) [inline pkg/docker/containers-view.jsx]: This needs to be the other way around. The arguments to ngettext should be the strings, and ```cockpit.format``` should operate on the result. At least that's how I understand it. If this does actually work, both the scanning and translating, then let me know.

File: pkg/docker/containers-view.jsx [hunk 6]
```diff
@@ -371,60 +417,98 @@ var ImageList = React.createClass({
...
+
+        var vulnerableInfo = this.state.vulnerableInfos[image.Id.replace(/^sha256:/, '')];
+        if (vulnerableInfo) {
+            var count = vulnerableInfo.vulnerabilities.length;
+            if (count > 0)
+                vulnerabilityColumn = (
+                    <div>
+                        <span className="pficon pficon-warning-triangle-o"></span>
+                        &nbsp;
+                        { cockpit.ngettext('1 Vulnerability', cockpit.format('$0 Vulnerabilities', count), count) }
```
Comment (stefwalter) [inline pkg/docker/containers-view.jsx]: This needs to be the other way around. The arguments to ngettext should be the strings, and ```cockpit.format``` should operate on the result. At least that's how I understand it. If this does actually work, both the scanning and translating, then let me know.

File: pkg/docker/docker.css [hunk 7]
```diff
@@ -1,3 +1,18 @@
+@import "/plot.css";
+@import "/term.css";
```
Comment (stefwalter) [inline pkg/docker/docker.css]: Out of interest, why were these two lines added here? Not necessarily a problem, but seem to be unrelated to this pull request.

**Comment by stefwalter** [comment]
This is also missing documentation in:

```
doc/guide/feature-docker.xml
```

In particular the documentation for me to be able to test this should be similar to the documentation found in that file about the feature, what it calls to do the scans etc.

File: pkg/docker/atomic.js (hunk 2 above)
Comment (larskarlitski) [inline pkg/docker/atomic.js]: Indeed, thanks for catching this. It's a remnant of a few refactorings.

File: pkg/docker/containers-view.jsx (hunk 3 above)
Comment (larskarlitski) [inline pkg/docker/containers-view.jsx]: Yes it is. Docker gives us an Object instead of an array there. I've separated this change into its own commit.

File: pkg/docker/containers-view.jsx (hunk 4 above)
Comment (larskarlitski) [inline pkg/docker/containers-view.jsx]: Thanks, fixed.

File: pkg/docker/docker.css (hunk 7 above)
Comment (larskarlitski) [inline pkg/docker/docker.css]: They are indeed unnecessary.

**Comment by stefwalter** [comment]
There's a consistent failure like this:

```
bash: /var/lib/atomic/scan_summary.json: No such file or directory
CalledProcessError: Command 'echo '{"e02e811dd08fd49e7f6032625495118e63f597eb150403d02e3238af1df240ba": {"UUID": "e02e811dd08fd49e7f6032625495118e63f597eb150403d02e3238af1df240ba", "Scan Type": "Standards Compliance", "json_file": "/var/lib/atomic/atomic_scan_openscap/2016-12-11-12-09-57-674487/e02e811dd08fd49e7f6032625495118e63f597eb150403d02e3238af1df240ba/json", "Time": "2016-12-11T12:09:59", "Vulnerable": true, "Scanner": "openscap"}}' > /var/lib/atomic/scan_summary.json' returned non-zero exit status 1
```

File: pkg/docker/containers-view.jsx [hunk 8]
```diff
@@ -308,6 +309,50 @@ var ImageDetails = React.createClass({
...
+
+        } else if (info.vulnerabilities.length === 0) {
+            text = _('The scan from $time ($type) found no vulnerabilities.');
+
+        } else {
+            text = cockpit.ngettext('The scan from $time ($type) found one vulnerability:',
+                                    'The scan from $time ($type) found $count vulnerabilities:',
+                                    info.vulnerabilities.length);
+
+            var rows = info.vulnerabilities.map(function (vulnerability) {
```
Comment (stefwalter) [inline pkg/docker/containers-view.jsx]: This relies on variable hoisting and is confusing. I would suggest declaring the variable outside the if block.

Comment (stefwalter) [inline pkg/docker/containers-view.jsx]: Hmmm, looks like you have declared the variable above. I would just remove the 'var ' while merging ...

**Comment by stefwalter** [comment]
I wanted to help fix this and merge it but couldn't get the tests to pass. 

 * I see this on Fedora 25: ```Error: #containers-images tr:contains("busybox:latest") td:contains("2 Vulnerabilities") not found```
 * No 'Security' tab when running Fedora 25
 * It seems like the test needs to be skipped on many other operating systems. See other tests for how to do that.

File: test/verify/check-docker [hunk 9]
```diff
@@ -498,5 +498,68 @@ CMD ["/bin/sh"]
...
         b.wait_in_text("#container-details-cpu-row", "1024 shares")
 
+class TestAtomicScan(MachineCase):
+    def testBasic(self):
+        b = self.browser
+        m = self.machine
+
+        m.execute("systemctl start docker")
+
+        # HACK put scan results for one of the images into /var/lib/atomic
```
Comment (dperpeet) [inline test/verify/check-docker]: Is there a bug/link to go with this `HACK`? So we know when we can change this / remove the hack.

File: pkg/docker/containers-view.jsx [hunk 10]
```diff
@@ -319,7 +364,8 @@ var ImageList = React.createClass({
     getInitialState: function () {
         return {
             images: [],
-            pulling: []
+            pulling: [],
+            vulnerableInfos: {}
```
Comment (dperpeet) [inline pkg/docker/containers-view.jsx]: Technically it's against our custom to have this info in the component state: This is information we'll probably also need to consume in other places (e.g. `X vulnerabilities found in Y images`). We can leave as is for now, but it'll probably need refactoring in the near future.

**Comment by larskarlitski** [comment]
> I see this on Fedora 25: Error: #containers-images tr:contains("busybox:latest") td:contains("2 Vulnerabilities") not found

I didn't see this because of a race: `atomic_dbus.py` gets activated on the `org.atomic` name, but only very briefly does it actually hold on to the name. Cockpit didn't catch the `NameOwnerChanged` signal in time in all cases. If it did, it closed the channel immediately.

I worked around this by making cockpit not track the name. Thanks to @stefwalter for the idea.

> No 'Security' tab when running Fedora 25

I guess this is because of that same race. Can you please try again with the latest patch?

> It seems like the test needs to be skipped on many other operating systems. See other tests for how to do that.

Added skips for all systems in `test/README` that don't have atomic.

File: test/verify/check-docker (hunk 9 above)
Comment (larskarlitski) [inline test/verify/check-docker]: No. It's hacky not because it's working around some issue, but because I don't yet know how we can do this in a better way. We should probably make a known-vulnerable image and scan it on every test?

File: pkg/docker/containers-view.jsx (hunk 10 above)
Comment (larskarlitski) [inline pkg/docker/containers-view.jsx]: I don't understand. Where else would you put it?

Comment (dperpeet) [inline pkg/docker/containers-view.jsx]: Usually we would separate the data store from the view, so have more data in the props. But I definitely don't think it's worth changing this now. We can have a general discussion about this sometime in the near future and consider whether your approach should become the default.

**Comment by larskarlitski** [comment]
The test failure on the atomics was due to `/var/lib/atomic` missing. Just pushed a fix for that.

Note that they will still fail, because the D-Bus service doesn't work at all on atomic right now.

**Comment by stefwalter** [comment]
> Note that they will still fail, because the D-Bus service doesn't work at all on atomic right now.

Could you include a HACK that fixes it. I think you said it's a one line python change. Otherwise we have no concrete idea this actually works once the atomic guys do a release in the next couple days.

**Comment by larskarlitski** [comment]
> Could you include a HACK that fixes it. I think you said it's a one line python change. Otherwise we have no concrete idea this actually works once the atomic guys do a release in the next couple days.

No, that's a separate issue that I worked around with not tracking the name from cockpit.

The problem here is that the atomic is not installed correctly. `python-slip-dbus` is missing.

**Comment by dperpeet** [comment]
fedora-atomic and rhel-atomic seem to fail the new test with the same error:
```
# testBasic (check_docker.TestAtomicScan)
#
Resource Error: Connection refused http://10.111.118.102:9090/docker
Restarting browser...
Wrote TestAtomicScan-testBasic-FAIL.png
Journal extracted to TestAtomicScan-testBasic-10.111.118.102-FAIL.log
not ok 81 testBasic (check_docker.TestAtomicScan) duration: 143s
Traceback (most recent call last):
  File "./verify/check-docker", line 567, in testBasic
    b.wait_present(row_selector + ' td span.pficon-warning-triangle-o')
  File "/build/cockpit/test/common/testlib.py", line 220, in wait_present
    return self.wait_js_func('ph_is_present', selector)
  File "/build/cockpit/test/common/testlib.py", line 214, in wait_js_func
    return self.phantom.wait("%s(%s)" % (func, ','.join(map(jsquote, args))))
  File "/build/cockpit/test/common/testlib.py", line 721, in <lambda>
    return lambda *args: self._invoke(name, *args)
  File "/build/cockpit/test/common/testlib.py", line 744, in _invoke
    raise Error(res['error'])
Error: timeout
```

**Comment by larskarlitski** [comment]
> fedora-atomic and rhel-atomic seem to fail the new test with the same error:

No, this is the error I mentioned above. The D-Bus interface doesn't work at all on the atomics, because there's a package missing. They are aware that this is an issue. Should we blacklist the test until then?

File: test/verify/check-docker (hunk 9 above)
Comment (larskarlitski) [inline test/verify/check-docker]: Good point, thanks. I've changed the comment.

**Comment by dperpeet** [comment]
I can't see the security results as a regular user, I need to check the box on the login page. Can this be escalated automatically?
If it's not easily solvable, I think this is ok to do as a follow-up and shouldn't block the main part going in.

**Comment by dperpeet** [comment]
for reference: to test this on fedora-25, I added an entry to the scans array in `/etc/atomic.d/atomic_scan_openscap`:
```
      { name: standards_compliance,
        args: ['oscapd-evaluate', 'scan', '--targets', 'chroots-in-dir:///scanin',  '--output', '/scanout', '--no-cve-scan', '-j1'],
        description: "Performs a standard scan"
      }
```
then I ran a scan with `atomic scan --scanner atomic_scan_openscap --scan_type standards_compliance docker.io/fedora:24`

**Comment by dperpeet** [comment]
> I can't see the security results as a regular user, I need to check the box on the login page. Can this be escalated automatically?
> If it's not easily solvable, I think this is ok to do as a follow-up and shouldn't block the main part going in.

This pull request looks good, I filed two follow-up issues. If this privilege issue is easily solved then let's get that in, otherwise I'm ok with a follow-up on that also.

**Comment by dperpeet** [comment]
Discussion on irc led to the decision to look at the privilege question as a follow-up.
I posted a fix-up pull request in #5593 with the skipped tests on rhel-atomic and fedora-atomic
